File: backend/app/admin/schemas/node_group.py
from datetime import datetime
from typing import Optional, Literal

from pydantic import Field, model_validator

from ._base import CamelModel


# --- Attr Strategy ---

class AttrStrategyItem(CamelModel):
    field_key: str = Field(..., min_length=1)
    strategy: Literal["fixed", "random", "increment", "range"]
    fixed_value: Optional[str] = None
    pool: Optional[list[str]] = None
    base: Optional[str] = None
    step: Optional[str] = None
    min: Optional[int] = None
    max: Optional[int] = None

    @model_validator(mode="after")
    def validate_strategy_params(self):
        if self.strategy == "fixed":
            if not self.fixed_value:
                raise ValueError("fixed 策略必须提供 fixedValue")
        elif self.strategy == "random":
            if not self.pool or len(self.pool) < 1:
                raise ValueError("random 策略必须提供 pool 且至少包含 1 个值")
        elif self.strategy == "increment":
            if self.base is None or self.step is None:
                raise ValueError("increment 策略必须提供 base 和 step")
        elif self.strategy == "range":
            if self.min is None or self.max is None:
                raise ValueError("range 策略必须提供 min 和 max")
        return self


# --- Edge Strategy ---

class EdgeStrategyItem(CamelModel):
    target_group_id: str = Field(..., min_length=1)
    edge_type_code: str = Field(..., min_length=1)
    mode: Literal["modulo", "one_to_n", "all_to_all", "dense"]
    ratio_k: Optional[int] = None

    @model_validator(mode="after")
    def validate_mode_params(self):
        if self.mode in ("modulo", "one_to_n"):
            if self.ratio_k is None or self.ratio_k <= 0:
                raise ValueError(f"{self.mode} 模式必须提供 ratioK 且 > 0")
        return self


# --- NodeGroup CRUD ---

class NodeGroupCreate(CamelModel):
    node_type_id: str = Field(..., min_length=1)
    group_name: str = Field(..., min_length=1, max_length=200)
    node_count: int = Field(..., ge=1, le=1_000_000)
    name_template: str = Field(default="{group}-{i:05d}", max_length=200)
    attr_strategies: list[AttrStrategyItem] = []
    edge_strategies: Optional[list[EdgeStrategyItem]] = None


class NodeGroupUpdate(CamelModel):
    group_name: Optional[str] = Field(default=None, min_length=1, max_length=200)
    node_count: Optional[int] = Field(default=None, ge=1, le=1_000_000)
    name_template: Optional[str] = None
    attr_strategies: Optional[list[AttrStrategyItem]] = None
    edge_strategies: Optional[list[EdgeStrategyItem]] = None


class NodeGroupItem(CamelModel):
    id: str
    topology_id: str
    node_type_id: str
    group_name: str
    node_count: int
    name_template: str
    attr_strategies: list[AttrStrategyItem]
    edge_strategies: Optional[list[EdgeStrategyItem]]
    is_materialized: bool
    created_at: datetime
    updated_at: datetime


# --- Group Graph (for canvas rendering) ---

class MacroNodeStatus(CamelModel):
    online: int = 0
    offline: int = 0


class MacroNode(CamelModel):
    id: str
    topology_id: str
    node_type_id: str
    group_name: str
    node_count: int
    is_materialized: bool
    status_breakdown: MacroNodeStatus
    x: Optional[float] = None
    y: Optional[float] = None


class MacroEdge(CamelModel):
    source_group_id: str
    target_group_id: str
    edge_type_code: str
    mode: str
    ratio_k: Optional[int] = None
    total_edge_count: int
    visual_source_is_macro: Optional[bool] = None


class GroupGraphData(CamelModel):
    macro_nodes: list[MacroNode]
    macro_edges: list[MacroEdge]

File: frontend/src/api/nodeGroup.ts
import { apiGet, apiPost, apiPut, apiPatch, apiDelete } from './http'

// ============ Attr Strategy ============

export interface AttrStrategyItem {
  fieldKey: string
  strategy: 'fixed' | 'random' | 'increment' | 'range'
  fixedValue?: string | null
  pool?: string[] | null
  base?: string | null
  step?: string | null
  min?: number | null
  max?: number | null
}

// ============ Edge Strategy ============

export interface EdgeStrategyItem {
  targetGroupId: string
  edgeTypeCode: string
  mode: 'modulo' | 'one_to_n' | 'all_to_all' | 'dense'
  ratioK?: number | null
}

// ============ NodeGroup CRUD ============

export interface NodeGroupCreate {
  nodeTypeId: string
  groupName: string
  nodeCount: number
  nameTemplate?: string
  attrStrategies?: AttrStrategyItem[]
  edgeStrategies?: EdgeStrategyItem[] | null
}

export interface NodeGroupUpdate {
  groupName?: string | null
  nodeCount?: number | null
  nameTemplate?: string | null
  attrStrategies?: AttrStrategyItem[] | null
  edgeStrategies?: EdgeStrategyItem[] | null
}

export interface NodeGroupItem {
  id: string
  topologyId: string
  nodeTypeId: string
  groupName: string
  nodeCount: number
  nameTemplate: string
  attrStrategies: AttrStrategyItem[]
  edgeStrategies: EdgeStrategyItem[] | null
  isMaterialized: boolean
  createdAt: string
  updatedAt: string
}

// ============ Materialize ============

export interface MaterializeProgress {
  phase: 'nodes' | 'edges' | 'done'
  current: number
  total: number
  pct: number
  elapsedMs: number
}

export interface MaterializeResult {
  materializedNodes: number
  materializedEdges: number
  elapsedMs: number
}

// ============ Group Graph ============

export interface MacroNodeStatus {
  online: number
  offline: number
}

export interface MacroNode {
  id: string
  topologyId: string
  nodeTypeId: string
  groupName: string
  nodeCount: number
  isMaterialized: boolean
  statusBreakdown: MacroNodeStatus
  x: number | null
  y: number | null
}

export interface MacroEdge {
  sourceGroupId: string
  targetGroupId: string
  edgeTypeCode: string
  mode: string
  ratioK: number | null
  totalEdgeCount: number
  visualSourceIsMacro: boolean | null
}

export interface GroupGraphData {
  macroNodes: MacroNode[]
  macroEdges: MacroEdge[]
}

// ============ Node Group Alarms ============

export interface NodeGroupAlarmItem {
  id: string
  nodeGroupId: string
  alarmIndex: number
  attrs: Record<string, string | null>
  createdAt: string
  updatedAt: string
}

export interface NodeGroupAlarmCreate {
  attrs?: Record<string, string | null>
}

export interface NodeGroupAlarmAttrSet {
  attrs: Record<string, string | null>
}

// ============ API Functions ============

export const nodeGroupApi = {
  list: (topologyId: string): Promise<{ items: NodeGroupItem[] }> =>
    apiGet(`/topologies/${topologyId}/node-groups`),

  get: (id: string): Promise<NodeGroupItem> =>
    apiGet(`/node-groups/${id}`),

  create: (topologyId: string, data: NodeGroupCreate): Promise<{ id: string }> =>
    apiPost(`/topologies/${topologyId}/node-groups`, data),

  update: (id: string, data: NodeGroupUpdate): Promise<{ id: string }> =>
    apiPut(`/node-groups/${id}`, data),

  delete: (id: string): Promise<{ id: string }> =>
    apiDelete(`/node-groups/${id}`),

  materialize: (id: string): Promise<MaterializeResult> =>
    apiPost(`/node-groups/${id}/materialize`),

  getGroupGraph: (topologyId: string): Promise<GroupGraphData> =>
    apiGet(`/topologies/${topologyId}/group-graph`),

  updatePosition: (id: string, data: { x: number; y: number }): Promise<{ id: string; x: number; y: number }> =>
    apiPatch(`/node-groups/${id}/position`, data),

  listAlarms: (groupId: string): Promise<NodeGroupAlarmItem[]> =>
    apiGet(`/node-groups/${groupId}/alarms`),

  createAlarm: (groupId: string, data: NodeGroupAlarmCreate = {}): Promise<NodeGroupAlarmItem> =>
    apiPost(`/node-groups/${groupId}/alarms`, data),

  updateAlarmAttrs: (alarmId: string, data: NodeGroupAlarmAttrSet): Promise<NodeGroupAlarmItem> =>
    apiPut(`/node-group-alarms/${alarmId}/attrs`, data),

  deleteAlarm: (alarmId: string): Promise<null> =>
    apiDelete(`/node-group-alarms/${alarmId}`),
}
